File: solutions/typescript/phone-number/1/phone-number.ts
function assert_original_valid(phone: string): void {
  if ( phone.match(/[A-Za-z]/) ) throw('Letters not permitted');
  if ( phone.match(/[@:!]/) ) throw('Punctuations not permitted');
}

function sanitize_cleaned(cleaned: string): string {
  if ( cleaned.length === 11 && cleaned[0] !== '1' ) throw('11 digits must start with 1');
  if ( cleaned.length === 11 ) cleaned = cleaned.substring(1);
  if ( cleaned.length > 10 ) throw('More than 11 digits');
  if ( cleaned.length !== 10 ) throw('Incorrect number of digits');
  if ( cleaned[0] === '0' ) throw('Area code cannot start with zero');
  if ( cleaned[0] === '1' ) throw('Area code cannot start with one');
  if ( cleaned[3] === '0' ) throw('Exchange code cannot start with zero');
  if ( cleaned[3] === '1' ) throw('Exchange code cannot start with one');
  return cleaned;
}
export function clean(phone: string): string {
  assert_original_valid(phone);
  return sanitize_cleaned(phone.replace(/[^0-9]/g, ''));
}

File: solutions/elixir/phone-number/1/lib/phone_number.ex
defmodule PhoneNumber do
  @doc """
  Remove formatting from a phone number if the given number is valid. Return an error otherwise.
  """
  @spec clean(String.t()) :: {:ok, String.t()} | {:error, String.t()}
  def clean(raw) do
    if String.match?(raw, ~r/[^0-9\.+\- \(\)]/) do
      {:error, "must contain digits only"}
    else
      cleaned = raw
      |> String.replace(~r/\D/, "")
      cleaned = if String.match?(cleaned, ~r/^1\d{10}/) do
        String.slice(cleaned, 1..-1)
      else
        cleaned
      end
      cond do
        String.length(cleaned) == 11 ->
          {:error, "11 digits must start with 1"}
        String.match?(cleaned, ~r/0\d{9}/) ->
          {:error, "area code cannot start with zero"}
        String.length(cleaned) != 10 ->
          {:error, "incorrect number of digits"}
        String.match?(cleaned, ~r/1\d{9}/) ->
          {:error, "area code cannot start with one"}
        String.match?(cleaned, ~r/[2-9]\d\d0\d{6}/) ->
          {:error, "exchange code cannot start with zero"}
        String.match?(cleaned, ~r/[2-9]\d\d1\d{6}/) ->
          {:error, "exchange code cannot start with one"}
        String.match?(cleaned, ~r/[2-9]\d\d[2-9]\d{6}/) ->
          {:ok, cleaned}
        true ->
        {:error, "incorrect number of digits"}
      end
    end
  end
end
